File: front-end/src/lib/types/generation.ts
export type QuestionType =
  | "MCQ"
  | "ASSERTION_REASON"
  | "VSA"
  | "SA"
  | "LA"
  | "CASE_STUDY"
  | "FILL_IN_THE_BLANK"
  | "TRUE_FALSE"
  | "OTHER"

export type BloomsLevel =
  | "REMEMBERING"
  | "UNDERSTANDING"
  | "APPLYING"
  | "ANALYSING"
  | "EVALUATING"
  | "CREATING"

export type ChapterAllocation = {
  chapter_number: number | null
  chapter_name: string
  question_count: number
  blooms_level: BloomsLevel | null
  has_internal_choice: boolean
  is_assertion_reason: boolean
  marks?: number | null
}

export type BlueprintSection = {
  section_name: string
  question_type: QuestionType
  marks_each: number
  chapter_allocations: ChapterAllocation[]
  section_instructions: string | null
  section_total_marks?: number
}

export type QuestionPaperBlueprint = {
  school_name: string | null
  exam_title: string | null
  subject: string
  grade: number
  total_marks: number
  duration_minutes: number | null
  exam_date: string | null
  general_instructions: string[]
  sections: BlueprintSection[]
  blooms_targets: Partial<Record<BloomsLevel, number>> | null
  allocated_marks?: number
}

export type SampleBlueprintSummary = {
  id: string
  slug: string
  label: string
  total_marks: number
}

export type SampleBlueprintDetail = SampleBlueprintSummary & {
  blueprint: QuestionPaperBlueprint
}

export type SelectedChapterRef = {
  book_code: string
  chapter_number: number
  chapter_name: string
}

export type GeneratePaperPayload = {
  blueprint: QuestionPaperBlueprint
  selected_chapters: SelectedChapterRef[]
  subject: string
  grade: number
  teacher_instructions?: string | null
}

export type GenerationResult = {
  blueprint: QuestionPaperBlueprint
  final_paper: Record<string, unknown>
  final_answer_key: Record<string, unknown>
  generated_items: Record<string, unknown>[]
}

export const QUESTION_TYPE_LABELS: Record<QuestionType, string> = {
  MCQ: "MCQ",
  ASSERTION_REASON: "Assertion-Reason",
  VSA: "VSA",
  SA: "SA",
  LA: "LA",
  CASE_STUDY: "CBQ",
  FILL_IN_THE_BLANK: "Fill in the Blank",
  TRUE_FALSE: "True / False",
  OTHER: "Other",
}

export const SECTION_QUESTION_TYPES: QuestionType[] = [
  "MCQ",
  "VSA",
  "SA",
  "LA",
  "CASE_STUDY",
]

export const DURATION_OPTIONS: { label: string; minutes: number }[] = [
  { label: "1 Hour", minutes: 60 },
  { label: "1.5 Hours", minutes: 90 },
  { label: "2 Hours", minutes: 120 },
  { label: "2.5 Hours", minutes: 150 },
  { label: "3 Hours", minutes: 180 },
]

export function sectionAllocatedMarks(section: BlueprintSection): number {
  return section.chapter_allocations.reduce(
    (sum, alloc) => sum + alloc.question_count * (alloc.marks ?? section.marks_each),
    0
  )
}

export function blueprintAllocatedMarks(blueprint: QuestionPaperBlueprint): number {
  return blueprint.sections.reduce(
    (sum, section) => sum + sectionAllocatedMarks(section),
    0
  )
}

export function classGradeToNumber(classGrade: string | null): number {
  const match = classGrade?.match(/\d+/)
  return match ? Number(match[0]) : 10
}

export function sectionLetter(index: number): string {
  return String.fromCharCode(65 + (index % 26))
}

function normalizeName(name: string): string {
  return name.toLowerCase().replace(/\s+/g, " ").trim()
}

function matchSelectedChapter(
  alloc: ChapterAllocation,
  selected: SelectedChapterRef[]
): SelectedChapterRef | null {
  if (alloc.chapter_number != null) {
    const byNumber = selected.find((ch) => ch.chapter_number === alloc.chapter_number)
    if (byNumber) return byNumber
  }

  const needle = normalizeName(alloc.chapter_name)
  return (
    selected.find((ch) => {
      const name = normalizeName(ch.chapter_name)
      return name === needle || name.includes(needle) || needle.includes(name)
    }) ?? null
  )
}

/** Keep only allocations that map onto the notebook's selected chapters. */
export function rematchBlueprintChapters(
  blueprint: QuestionPaperBlueprint,
  selectedChapters: SelectedChapterRef[]
): QuestionPaperBlueprint {
  return {
    ...blueprint,
    sections: blueprint.sections.map((section) => ({
      ...section,
      chapter_allocations: section.chapter_allocations.flatMap((alloc) => {
        const matched = matchSelectedChapter(alloc, selectedChapters)
        if (!matched) return []
        return [
          {
            ...alloc,
            chapter_number: matched.chapter_number,
            chapter_name: matched.chapter_name,
          },
        ]
      }),
    })),
  }
}

export function hasForeignChapterAllocations(
  blueprint: QuestionPaperBlueprint,
  selectedChapters: SelectedChapterRef[]
): boolean {
  const allowed = new Set(selectedChapters.map((ch) => ch.chapter_number))
  return blueprint.sections.some((section) =>
    section.chapter_allocations.some(
      (alloc) =>
        alloc.chapter_number == null || !allowed.has(alloc.chapter_number)
    )
  )
}

export function emptyBlueprint(partial?: Partial<QuestionPaperBlueprint>): QuestionPaperBlueprint {
  return {
    school_name: null,
    exam_title: null,
    subject: "Mathematics",
    grade: 10,
    total_marks: 40,
    duration_minutes: 90,
    exam_date: null,
    general_instructions: [],
    sections: [],
    blooms_targets: null,
    ...partial,
  }
}

File: backend/app/schemas/generation.py
from datetime import date
from enum import Enum
from typing import Self, TypedDict
from uuid import UUID

from pydantic import BaseModel, Field, computed_field, model_validator

from app.schemas.notebook import SelectedChapter


class QuestionType(str, Enum):
    MCQ = "MCQ"
    ASSERTION_REASON = "ASSERTION_REASON"
    VSA = "VSA"
    SA = "SA"
    LA = "LA"
    CASE_STUDY = "CASE_STUDY"
    FILL_IN_THE_BLANK = "FILL_IN_THE_BLANK"
    TRUE_FALSE = "TRUE_FALSE"
    OTHER = "OTHER"


class BloomsLevel(str, Enum):
    REMEMBERING = "REMEMBERING"
    UNDERSTANDING = "UNDERSTANDING"
    APPLYING = "APPLYING"
    ANALYSING = "ANALYSING"
    EVALUATING = "EVALUATING"
    CREATING = "CREATING"


class ChapterAllocation(BaseModel):
    chapter_number: int | None = Field(
        default=None,
        description="Catalog chapter number when known; null until rematched to notebook chapters.",
    )
    chapter_name: str = Field(description="Lesson / chapter display name.")
    question_count: int = Field(ge=1, description="Number of questions for this chapter in the section.")
    blooms_level: BloomsLevel | None = Field(
        default=None,
        description="Optional Bloom's level for generation; not shown as a form section.",
    )
    has_internal_choice: bool = False
    is_assertion_reason: bool = Field(
        default=False,
        description="True for Assertion-Reason items nested under an MCQ section.",
    )
    marks: float | None = Field(
        default=None,
        description="Optional per-allocation marks override; defaults to section marks_each.",
    )


class BlueprintSection(BaseModel):
    section_name: str = Field(
        description="Display name for the section, e.g. 'MCQ', 'VSA', 'CBQ'."
    )
    question_type: QuestionType = Field(
        description="Question type for this section (CBQ maps to CASE_STUDY)."
    )
    marks_each: float = Field(description="Marks per question in this section.")
    chapter_allocations: list[ChapterAllocation] = Field(
        default_factory=list,
        description="Per-chapter question counts within this section.",
    )
    section_instructions: str | None = Field(
        default=None,
        description="Optional notes at the section header.",
    )

    @computed_field  # type: ignore[prop-decorator]
    @property
    def section_total_marks(self) -> float:
        return sum(
            a.question_count * (a.marks if a.marks is not None else self.marks_each)
            for a in self.chapter_allocations
        )


class QuestionPaperBlueprint(BaseModel):
    school_name: str | None = Field(
        default=None,
        description="Institution name; null if absent.",
    )
    exam_title: str | None = Field(
        default=None,
        description="Exam / sheet title (e.g. 'Pre-Mid Term Examination').",
    )
    subject: str = Field(description="Subject domain, e.g. 'Mathematics'.")
    grade: int = Field(description="Class / grade level, e.g. 10.")
    total_marks: int = Field(description="Target total marks for the paper.")
    duration_minutes: int | None = Field(
        default=None,
        description="Duration in minutes (e.g. '2 Hours' -> 120); null if absent.",
    )
    exam_date: date | None = Field(
        default=None,
        description="Scheduled exam date if set.",
    )
    general_instructions: list[str] = Field(
        default_factory=list,
        description=(
            "Editable General Instructions template items included in the final paper "
            "and question-generation prompt."
        ),
    )
    sections: list[BlueprintSection] = Field(
        description="Ordered question-type sections making up the paper.",
    )
    blooms_targets: dict[BloomsLevel, float] | None = Field(
        default=None,
        description="Optional Bloom's mark targets from the sample template.",
    )

    @computed_field  # type: ignore[prop-decorator]
    @property
    def allocated_marks(self) -> float:
        return sum(section.section_total_marks for section in self.sections)

    @model_validator(mode="after")
    def validate_allocated_marks(self) -> Self:
        if abs(self.allocated_marks - self.total_marks) > 1e-9:
            raise ValueError(
                f"section allocations total {self.allocated_marks:g} marks, "
                f"but total_marks is {self.total_marks}"
            )
        return self


class SlotSubPart(BaseModel):
    label: str
    marks: float
    has_internal_choice: bool = False


class Slot(BaseModel):
    slot_id: str
    section_name: str
    question_number: int
    question_type: QuestionType
    marks: float
    chapter_number: int
    chapter_name: str
    book_code: str | None = None
    content_types: list[str] = Field(default_factory=list)
    has_internal_choice: bool = False
    blooms_level: BloomsLevel | None = None
    sub_parts: list[SlotSubPart] = Field(default_factory=list)


class RubricStep(BaseModel):
    description: str
    marks: float


class GeneratedQuestion(BaseModel):
    slot_id: str = Field(description="Must exactly match the slot_id given in the question spec.")
    question_text: str = Field(
        description="For ASSERTION_REASON: only Assertion (A) and Reason (R) — options are appended later."
    )
    options: list[str] | None = Field(
        default=None,
        description="Exactly 4 options if question_type is MCQ, else null",
    )
    correct_option: str | None = Field(
        default=None,
        description="'a'/'b'/'c'/'d' if MCQ or ASSERTION_REASON, else null",
    )
    answer: str
    marking_rubric: list[RubricStep] = Field(
        description="Steps whose marks sum exactly to the required marks"
    )
    alternate_question_text: str | None = Field(
        default=None,
        description="Only if has_internal_choice is true.",
    )
    alternate_answer: str | None = None
    source_chunk_ids: list[str] = Field(default_factory=list)


class GeneratedPaperResponse(BaseModel):
    items: list[GeneratedQuestion]


class SampleBlueprintSummary(BaseModel):
    id: UUID
    slug: str
    label: str
    total_marks: int


class SampleBlueprintDetail(SampleBlueprintSummary):
    blueprint: QuestionPaperBlueprint


class GeneratePaperRequest(BaseModel):
    """Request body for question paper generation."""

    blueprint: QuestionPaperBlueprint
    selected_chapters: list[SelectedChapter]
    subject: str
    grade: int
    teacher_instructions: str | None = None


class GenerationResult(BaseModel):
    blueprint: QuestionPaperBlueprint
    final_paper: dict
    final_answer_key: dict
    generated_items: list[dict]
    format_reference_path: str
    format_reference_is_default: bool = True


class GenerationState(TypedDict):
    question_paper: dict
    selected_chapters: list[dict]
    subject: str
    grade: int
    teacher_instructions: str | None
    format_reference_path: str | None
    slots: list[dict]
    generated_items: list[dict]
    final_paper: dict | None
    final_answer_key: dict | None
